Dockerfile

```diff
@@ -1,11 +1,35 @@
-FROM alpine:3.6 as certs
+# base golang image
+ARG GOVER="1.12.4-alpine3.9"
+FROM golang:${GOVER} as golang
 
-RUN apk add -U --no-cache ca-certificates
+ARG REPO
+ARG OS
+ARG ARCH
```

Comment:
Would advise against this kind of parameterization. This kind of stuff makes a lot of sense for a `Makefile` or bash script where you want more control and omnibus functionality.

For docker thats a bit of an anti-pattern, `ARG` itself was even a contested feature cause it can hinder idempotency for builds. The benefit of the approach in #4 is that anyone can clone the repo and build the dockerfile without having to figure out how to build the dockerfile. `docker build .` will produce a working image.

Reply:
Then how would you cross-build?

> `docker build .` will produce a working image

Yes, but only on the platform on which it runs. If you want anything else (windows/darwin/arm/ppc/s390x, amd64/arm64/riscv), you are depending on running it there.

Reply:
Also, how would you run things like `go fmt`, `go vet`, `go test`, etc.?

Reply:
> Yes, but only on the platform on which it runs.
AFAIK any time your running a docker image its always running on some thin linux vm. In the boot2docker days it used to be more apparent, but nowadays with "Docker for Mac" & "Docker for Windows" I believe its all https://github.com/linuxkit/linuxkit
Ref: https://blog.docker.com/2017/04/introducing-linuxkit-container-os-toolkit/
Ref: https://docs.docker.com/docker-for-mac/release-notes/

Building exclusively for linux should be appropriate for use with scratch images. For an ASG roller, I don't think the other architectures make sense either, you're more than likely going to run this on some amd64 ec2.

> Also, how would you run things like go fmt, go vet, go test, etc.?

Not sure I follow, would expect `go fmt` etc to be run outside the context of building an image.

Reply:
In the context of this, exclusively for Linux is _probably_ but not definite. I actually have run it on my MacBook pointing to AWS. And AWS has `arm64` instances as well, and I can run a k8s cluster entirely of Windows. Other OSes and archs either work or are on the way.

Reply:
I did a lot of porting of various pieces of core software to arm64, sometimes ppc64le and even some a bit of s390x. :-)

Reply:
See other comment thread, misunderstood that you're actually trying to use docker to build the binaries for use outside of docker. Was only trying to make the case that if you're running it in docker the os is going to be linux AFAIK.

Reply:
The OS _probably_ (not definitely) is going to be linux for docker, but the arch definitely can be one of several.

And, yes, as a pattern, I like to support binaries that can run outside of images.